l2geth/core/vm/evm.go

```diff
@@ -325,40 +371,47 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 
 	if UsingOVM {
 		// OVM_ENABLED
-
 		if evm.depth == 0 {
 			// We're back at the root-level message call, so we'll need to modify the return data
 			// sent to us by the OVM_ExecutionManager to instead be the intended return data.
 
-			if len(ret) >= 96 {
-				// We expect that EOA contracts return at least 96 bytes of data, where the first
-				// 32 bytes are the boolean success value and the next 64 bytes are unnecessary
-				// ABI encoding data. The actual return data starts at the 96th byte and can be
-				// empty.
-				success := ret[:32]
-				ret = ret[96:]
-
-				if !bytes.Equal(success, AbiBytesTrue) && !bytes.Equal(success, AbiBytesFalse) {
-					// If the first 32 bytes not either are the ABI encoding of "true" or "false",
-					// then the user hasn't correctly ABI encoded the result. We return the null
-					// hex string as a default here (an annoying default that would convince most
-					// people to just use the standard form).
-					ret = common.FromHex("0x")
-				} else if bytes.Equal(success, AbiBytesFalse) {
-					// If the first 32 bytes are the ABI encoding of "false", then we need to add an
-					// artificial error that represents the revert.
-					err = errExecutionReverted
+			// Attempt to decode the returndata as as ExecutionManager.run when
+			// it is not an `eth_call` and as ExecutionManager.simulateMessage
+			// when it is an `eth_call`. If the data is not decodable as ABI
+			// encoded bytes, then return nothing. If the data is able to be
+			// decoded as bytes, then attempt to decode as (bool, bytes)
+			isDecodable := true
+			run := runReturnData{}
+			switch evm.Context.EthCallSender {
+			case nil:
+				if err := evm.Context.OvmExecutionManager.ABI.Unpack(&run, "run", ret); err != nil {
+					isDecodable = false
+				}
+			default:
+				if err := evm.Context.OvmExecutionManager.ABI.Unpack(&run, "simulateMessage", ret); err != nil {
+					isDecodable = false
 				}
-			} else {
-				// User hasn't conformed the standard format, just return "null" for the success
-				// (with no return data) to convince them to use the standard.
-				ret = common.FromHex("0x")
 			}
 
-			if evm.Context.EthCallSender == nil {
-				log.Debug("Reached the end of an OVM execution", "ID", evm.Id, "Return Data", hexutil.Encode(ret), "Error", err)
+			switch isDecodable {
+			case true:
+				inner := innerData{}
+				// If this fails to decode, the nil values will be set in
+				// `inner`, meaning that it will be interpreted as reverted
+				// execution with empty returndata
+				_ = codec.Unpack(&inner, "call", run.ReturnData)
```

Comment:
Should maybe handle the error anyway?

Reply:
We want to be sure to not shadow `err` here so it would have to be:

```golang
if err := codec.Unpack(&inner, "call", run.ReturnData); err != nil {
    ...
}
```

We don't need to log the error and we don't want to return early if there is an error, any suggestions in particular?